compiler/rustc_metadata/src/fs.rs

```diff
@@ -0,0 +1,129 @@
+    }
+
+    out_filename
+}
+
+pub fn encode_and_write_metadata(
+    tcx: TyCtxt<'_>,
+    outputs: &OutputFilenames,
+) -> (EncodedMetadata, bool) {
+    #[derive(PartialEq, Eq, PartialOrd, Ord)]
+    enum MetadataKind {
+        None,
+        Uncompressed,
+        Compressed,
+    }
+
+    let metadata_kind = tcx
+        .sess
+        .crate_types()
+        .iter()
+        .map(|ty| match *ty {
+            CrateType::Executable | CrateType::Staticlib | CrateType::Cdylib => MetadataKind::None,
+
+            CrateType::Rlib => MetadataKind::Uncompressed,
+
+            CrateType::Dylib | CrateType::ProcMacro => MetadataKind::Compressed,
+        })
+        .max()
+        .unwrap_or(MetadataKind::None);
+
+    let crate_name = tcx.crate_name(LOCAL_CRATE);
+    let out_filename = filename_for_metadata(tcx.sess, crate_name.as_str(), outputs);
+    // To avoid races with another rustc process scanning the output directory,
+    // we need to write the file somewhere else and atomically move it to its
+    // final destination, with an `fs::rename` call. In order for the rename to
+    // always succeed, the temporary file needs to be on the same filesystem,
+    // which is why we create it inside the output directory specifically.
+    let metadata_tmpdir = TempFileBuilder::new()
+        .prefix("rmeta")
+        .tempdir_in(out_filename.parent().unwrap_or_else(|| Path::new("tmp")))
+        .unwrap_or_else(|err| tcx.sess.fatal(&format!("couldn't create a temp dir: {}", err)));
+    let metadata_tmpdir = MaybeTempDir::new(metadata_tmpdir, tcx.sess.opts.cg.save_temps);
+    let metadata_filename = metadata_tmpdir.as_ref().join(METADATA_FILENAME);
+    let metadata = match metadata_kind {
+        MetadataKind::None => {
+            if tcx.sess.opts.json_artifact_notifications {
+                tcx.sess
+                    .parse_sess
+                    .span_diagnostic
+                    .emit_artifact_notification(&out_filename, "metadata");
+            }
+            EncodedMetadata::empty()
+        }
+        MetadataKind::Uncompressed | MetadataKind::Compressed => {
+            encode_metadata(tcx, &metadata_filename);
+
+            let _prof_timer = tcx.sess.prof.generic_activity("write_crate_metadata");
+
+            let need_metadata_file = tcx.sess.opts.output_types.contains_key(&OutputType::Metadata);
+            let (metadata_filename, metadata_tmpdir) = if need_metadata_file {
+                if let Err(e) = non_durable_rename(&metadata_filename, &out_filename) {
+                    tcx.sess.fatal(&format!("failed to write {}: {}", out_filename.display(), e));
+                }
+                if tcx.sess.opts.json_artifact_notifications {
+                    tcx.sess
+                        .parse_sess
+                        .span_diagnostic
+                        .emit_artifact_notification(&out_filename, "metadata");
+                }
```

Comment:
More precisely, this function should pretty much do:
- create the file at `metadata_filename`;
- if `metadata_kind` is `Uncompressed | Compressed` fill it with data, otherwise keep it empty;
- if `need_metadata_file`, move `metadata_filename` to the correct location `out_filename` and emit the json notification;
- re-read either `metadata_filename` or `out_filename` with `EncodedMetadata::from_path`.


Does it make sense?

Reply:
> Could all the logic with need_metadata_file be moved out of this match? 

I'm sorry if I was wrong, it seems that your comment is based on the outdated commit. If that was wrong, could I ask you to tell me again how to fix the code based on my latest commit?